### ISSUE TITLE
Při importu dat z jiné instance ProArcu se zduplikují údaje o skenerech

### ISSUE THREAD

**Issue by svetlym** [proarc/proarc#1036]
Provozujeme ostrý a testovací ProArc. Importovali jsme teď několik titulů z testovacího do ostrého přes archivní formát a v nabídce skenerů v ostrém ProArcu máme teď některé zdvojené - stejné údaje, jiné device ID. Duplicitní skener navíc nejde ze seznamu odmazat, protože ID je přiřazeno k digitálnímu objektu. Předpokládám tedy, že se do seznamu přidal z importovaných dat.

Otázka je, co s tím. Správně asi měly skenery mít na ostrém i testovacím serveru stejná IDčka, je to ostatně ten samý skener. Idčko ale, předpokládám, generuje Proarc při zakládání záznamu o skeneru a nedá se nastavit ručně. A kromě toho dokumenty už jsou zpracované, takže na to je stejně pozdě. I tak by se to ale asi hodilo pro případné další instalace a pokud se to navíc propíše do existujících objektů, tak tím půjdou i sloučit různá idčka do jednoho.

• Navrhuju tedy možnost ručně změnit (případně sloučit) idčko zařízení. Bude to ovšem dost netriviální operace, protože to změní spoustu dat, která bude navíc následně třeba přeexportovat.

**Comment by luckajirku** [comment]
Tohle by měl ProArc umožňovat. Nejde jen o ostrý+testovací. My třeba máme dva ostré, používáme stejné skenery, ale ID se nám pak liší. A nebo skeny z jiných institucí, když chybějící část dostaneme odjinud a chtěli bychom tam mít (jak by to asi správně mělo být) správně nastavený skener - taky se to bude lišit. Potom to ID trochu postrádá smysl, když jedno a to samé zařízení musí mít v každém 
ProArcu odlišné ID.

**Comment by SykoraLukas** [comment]
Problém je, že objekt je identifikován UUID. Nepočítalo se s přenášením dat mezi ProArc aplikacemi. 

Postup by byl asi takový, že do metadat skeneru by bylo nutné zapsat další identifikátory, které se k danému zařízení vážou. Následně při importu archivního balíčku by se kontroloval nejen hlavní identifikátor UUID ale i ty další referenční.

**Comment by svetlym** [comment]
Nejjednodušší by zřejmě bylo, kdyby ten skenner měl stejný uuid v obou instancích proarcu. To ale problém nevyřeší zpětně. I když možná jo - pokud by šlo ručně změnit skenneru uuid, tak bych to mohl udělat i u toho duplicitního, kterej se vytvořit při importu dokumentů z druhý instance.

**Comment by SykoraLukas** [comment]
Ručně měnit uuid nepůjde.

**Comment by ZdenkaSera** [comment]
@SykoraLukas Prosím, jak je tohle nakonec udělané? Tj. pokud platí návrh řešení výše, které jsou další identifikátory, které při shodě zamezí duplikaci? Porovnávala jsem formulář zařízení na Inovatoce a u nás (3.7.3) a nezdá se, že by tam přibyla pole. Prosím tedy o radu, jak otestovat po nasazení nové verze k nám. Díky.

**Comment by SykoraLukas** [comment]
Je to vyřešeno v konfiguraci. Do zařízení se nakonec nedá nic přidat, protože v mixu na to neexistují relevantní pole.

proarc-common/src/main/resources/cz/cas/lib/proarc/common/config/proarc.properties

`## lists of devices IDs to include in configuration
devices=skener,audio

device.skener.mainUuid=uuid:12432423420
device.skener.otherUuid=uuid:12432423423,uuid:12432423422,uuid:12432423421`

**Comment by ZdenkaSera** [comment]
@SykoraLukas Aha - díky :-)

**Comment by ZdenkaSera** [comment]
@SykoraLukas Lukáši, prosím, mohl bys to přidat do dokumentace konfigu? Moc děkuji.

**Comment by ZdenkaSera** [comment]
@svetlym  Prosím, měli jste možnost nasadit u vás některou z verzí  od 3.8 a zkusit tuhle funkčnost? Děkuji.

**Comment by svetlym** [comment]
Verzi 3.8 nasazenou máme, k otestování jsme se ale ještě nedostali. Až to uděláme, dám sem vědět, jak to dopadlo.

**Comment by svetlym** [comment]
Dobrý den,

zkouším to ve verzi Verze: 4.0-dev a nedaří se mi skennery deduplikovat.

Idčka skennerů beru odtud:

![seznam_skenneru](https://user-images.githubusercontent.com/11628774/214061494-4938a0c4-27d4-4b09-b79c-eec4730c3a99.jpg)

Konfigurace vypadá takhle:

![konfigurace](https://user-images.githubusercontent.com/11628774/214061576-d2093b78-6b30-4f0a-ac58-fbfda79b6bfb.jpg)

Zkoušel jsem variantu s prefixem "uuid:" a "device:", v obou případech ale výsledek vypadá takhle:

![vyber_skenneru](https://user-images.githubusercontent.com/11628774/214061786-94c637b7-11a2-48ad-a3c2-d77688ec0693.jpg)

Čekal bych, že ve výběru uvidím Copybook a Suprascan II jen jednou.

Co dělám špatně?

**Comment by SykoraLukas** [comment]
Nevím už přesně jak se tohle dělalo, bylo to téměř před rokem. 

Ale myslím, že to bylo přizpůsobené tomu, aby nevznikaly další skenery, ne že se staré sloučí. Zjistím.

**Comment by SykoraLukas** [comment]
@ZdenkaSera to sice jo, ale jak se ptá @svetlym, je otázka jak to přesně je udělané.

a) přidám zařízení před importem a nevzniká nové zařízení, ale rovnou na importu se použije nově přiřazení uuid device z konfigurace. Nemá vliv na již existující zařízení.
b) přidám zařízení před importem a nevzniká nové zařízení, ale rovnou na importu se použije nově přiřazení uuid device z konfigurace. Při výběru z roletky jsou zařízení ovlivněny touto konfigurací. 

Myslím, že jsme to udělali podle varianty a) ale doplním omezení pro roletku aby to fungovalo jako b).

**Comment by ZdenkaSera** [comment]
@SykoraLukas Díky - přiznám se, že jsem o variantě b) vůbec neuvažovala, Ale jen teoretizuju, tohle jsme netestovali.

**Comment by svetlym** [comment]
> Myslím, že jsme to udělali podle varianty a) ale doplním omezení pro roletku aby to fungovalo jako b).

To by bylo určitě fajn, protože bez toho není jak poznat, že je to v konfiguraci nastavený správně.

**Comment by svetlym** [comment]
Zdravím,

nejsem si jistej, jestli se tady něco změnilo od chvíle, kdy jsem to posledně testoval. Má smysl zkoušet to znova?

**Comment by SykoraLukas** [comment]
Ano, je implemetováno. Vše je popsáno v historii tohoto issue.

**Comment by svetlym** [comment]
Updatnul jsem jádro na verzi 4.0.2 a nabídce mám pořád skennery dvakrát, a to ve starym i novym klientovi.

V konfiguraci to mám takhle:

# deduplikace zařízení
devices=Copibook, Suprascan_II
device.Copibook.mainUuid=uuid:56bff1b6-608c-4d23-91b6-62747676b9ff
device.Copibook.otherUuid=uuid:0466377c-f5f7-436b-8d09-4fabc5355255
device.Suprascan_II.mainUuid=uuid:94d5d6ab-0de6-476c-9813-f3097f2b0c36
device.Suprascan_II.otherUuid=uuid:015293d9-7106-4169-805d-3b1701493c10

**Comment by SykoraLukas** [comment]
Mohl byste přidat ještě výpiszařízení?

**Comment by svetlym** [comment]
![seznam_skenneru](https://github.com/proarc/proarc/assets/11628774/66186abc-7928-4a6d-9f3d-631900b4bd78)

**Comment by svetlym** [comment]
> Z výpisu je vidět, že id zařízení začínají prefixem `device:`, ale v konfiguraci máte `uuid:`. Prosím opravte a restartujte aplikaci.

Nepomohlo.

```
# deduplikace zařízení
devices=Copibook, Suprascan_II
device.Copibook.mainUuid=device:56bff1b6-608c-4d23-91b6-62747676b9ff
device.Copibook.otherUuid=device:0466377c-f5f7-436b-8d09-4fabc5355255
device.Suprascan_II.mainUuid=device:94d5d6ab-0de6-476c-9813-f3097f2b0c36
device.Suprascan_II.otherUuid=device:015293d9-7106-4169-805d-3b1701493c10
```

**Comment by ZdenkaSera** [comment]
@svetlym Pane Světlý z posledního zápisu se mi zdá, že tady je pořád problém. Pokud stále trvá, prosím, mohl byste vyměnit štítek "k testu" za štítek "chyba"? Takhle to unikne pozornosti vývojářů. Děkuji.

**Comment by svetlym** [comment]
@ZdenkaSera Ano, problém stále trvá. Na nastování labelů zřejmě nemám práva, nebo možná jen nevím, jak se to dělá. Nikde tu možnost nevidím.

**Comment by ZdenkaSera** [comment]
@svetlym Moc děkuji za reakci, štítky jsem přehodila.

**Comment by SykoraLukas** [comment]
Jakou verzi mate nasazenou?

**Comment by svetlym** [comment]
Aktuálně 4.2.3-SNAPSHOT, ale zkoušeli jsme to určitě na starší. Koukám teď to tý nový a zdá se, že to funguje. Ještě si to ale pro jistotu nechám potvrdit od kolegyní - jak s Proarcem nepracuju, tak si nejsem jistej, jestli jsem se proklikal na správný menu.

**Comment by svetlym** [comment]
Ve verzi 4.2.8 deduplikace funguje.
